{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import os\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "import gseapy as gp\n",
    "from matplotlib.patches import Patch\n",
    "import mygene\n",
    "import matplotlib.patheffects as PathEffects\n",
    "from adjustText import adjust_text\n",
    "import dask.dataframe as dd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the first step, the expected counts from each of the individual RSEM output files are combined together to create the combined count matrix for the whole dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def create_combined_count_matrix(folder_path):\n",
    "    \"\"\"\n",
    "    Create a combined count matrix from all the individual RSEM files in the specified folder.\n",
    "\n",
    "    Args:\n",
    "        folder_path (str): Path to the folder containing the RSEM files.\n",
    "\n",
    "    Returns:\n",
    "        pandas DataFrame: The combined count matrix for the dataset\n",
    "    \"\"\"\n",
    "    # empty list to store the individual count dataframes\n",
    "    count_dfs = []\n",
    "\n",
    "    # Iterate over all RSEM output files in the folder\n",
    "    for file_name in os.listdir(folder_path):\n",
    "       \n",
    "        # Generate the full path to access file\n",
    "        file_path = os.path.join(folder_path, file_name)\n",
    "        # Read the file into a Dask DataFrame, extract the gene_id and expected_count columns\n",
    "        count_df = dd.read_csv(file_path, sep='\\t', usecols=['gene_id', 'expected_count'])\n",
    "            \n",
    "        # Assign the sample name (derived from file name)\n",
    "        sample_name = file_name.split(\".\")[0]\n",
    "        count_df = count_df.rename(columns={'expected_count': sample_name})\n",
    "            \n",
    "        # change the index to gene_id column\n",
    "        count_df = count_df.set_index('gene_id')\n",
    "            \n",
    "        # Add the count data frame to the count_dfs list\n",
    "        count_dfs.append(count_df)\n",
    "\n",
    "    # Merge all individual count data frames based on same index (gene_id)\n",
    "    count_matrix = dd.concat(count_dfs, axis=1)\n",
    "    # Fill NaN values with 0 (in case of missing values)\n",
    "    count_matrix = count_matrix.fillna(0)\n",
    "    # Convert the expected counts to integer values which is necessary for DESeq2\n",
    "    count_matrix = count_matrix.round().astype(int)\n",
    "    # Compute the final DataFrame\n",
    "    count_matrix = count_matrix.compute()\n",
    "    #return the count matrix\n",
    "    return count_matrix\n",
    "\n",
    "# The folder path contating the RSEM files\n",
    "folder_path = 'GSE205099_RAW'\n",
    "# Generate the combined count matrix\n",
    "counts_df = create_combined_count_matrix(folder_path)\n",
    "# Save the count matrix\n",
    "counts_df.to_csv(f'{folder_path}_count_matrix.tsv', sep='\\t')\n",
    "print(\"Count matrix created successfully!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We convert the Ensembl ids to corresponding gene symbol. In case of multiple Ensembl IDs pointing to the same genesymbol, we added their counts together for the respective replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def process_counts_df(counts_df):\n",
    "    \"\"\"\n",
    "    To Process the counts dataframe by mapping the Ensembl IDs to Corresponding Gene Symbols and Perform additional transformations.\n",
    "\n",
    "    Args:\n",
    "        counts_df (pandas DataFrame): The counts matrix dataframe containing Ensembl IDs as index and gene counts as columns.\n",
    "\n",
    "    Returns:\n",
    "        counts_df: The processed counts matrix dataframe.\n",
    "    \"\"\"\n",
    "    # Initialize the mygene object\n",
    "    my_gene_object = mygene.MyGeneInfo()\n",
    "\n",
    "    # Extract the Ensembl IDs withour version numbers \n",
    "    ensembl_ids = counts_df.index.str.split('.').str[0].unique()\n",
    "\n",
    "    # Query the IDs to fetch the corresponding Gene Symbols\n",
    "    query_results = my_gene_object.querymany(ensembl_ids, scopes='ensembl.gene', fields='symbol', species='human')\n",
    "\n",
    "    # Using the Query result a dictionary is created to map Ensembl IDs to corresponding Gene Symbols\n",
    "    ensembl_to_symbol = {item['query']: item.get('symbol', 'NaN') for item in query_results}\n",
    "\n",
    "    # Add the correspodning Gene Symbols to the counts dataframe\n",
    "    counts_df['GeneSymbol'] = counts_df.index.str.split('.').str[0].map(ensembl_to_symbol)\n",
    "\n",
    "    # Change the index of the count matrix dataframe to the GeneSymbol column\n",
    "    counts_df.set_index('GeneSymbol', inplace=True)\n",
    "\n",
    "    # Transpose the dataframe\n",
    "    counts_df = counts_df.T\n",
    "\n",
    "    # Merge counts of the same but repeated gene symbol\n",
    "    counts_df = counts_df.groupby(level=0, axis=1).sum()\n",
    "\n",
    "    return counts_df\n",
    "\n",
    "\n",
    "counts_df=process_counts_df(counts_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we Create the metadata dataframe highlighting the biological condition of each replicate which is required for carrying out the Differential Expression Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assing_condition(replicate):\n",
    "    \"\"\"\n",
    "    Determines the condition of a given replicate based on its name.\n",
    "\n",
    "    Parameters:\n",
    "    replicate (str): The name of the replicate.\n",
    "\n",
    "    Returns:\n",
    "    str: The condition of the replicate. Either \"Covid\" or \"Control\".\n",
    "    \"\"\"\n",
    "    if replicate.split(\"_\")[1] == \"covid19\":\n",
    "        return \"Covid\"\n",
    "    else:\n",
    "        return \"Control\"\n",
    "\n",
    "# Extract the replicate name as a list\n",
    "replicates=list(counts_df.index)\n",
    "# Generate a list of the assocaited biological condition of each replicate     \n",
    "condition_list=[assing_condition(sample) for sample in replicates]\n",
    "# Generate the metadata \n",
    "metadata=pd.DataFrame(condition_list, index=replicates, columns=[\"Condition\"])\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now carry out the Differential Expression Analysis using DESeq2 by Comparing Infected Samples against Control "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out genes with low counts\n",
    "counts_df = counts_df.loc[:, counts_df.sum() >= 10]\n",
    "\n",
    "#Initialize the DESeq2 object\n",
    "dds = DeseqDataSet(\n",
    "    counts=counts_df,\n",
    "    metadata=metadata,\n",
    "    design_factors=\"Condition\",\n",
    "    n_cpus=32,  # n_cpus can be specified here or in the inference object\n",
    ")\n",
    "dds.deseq2()\n",
    "\n",
    "#Run the DESeq2 analysis, comparing the Covid against Control conditions\n",
    "stat_res = DeseqStats(dds, contrast=('Condition', 'Covid', 'Control'))\n",
    "#Summarize the results\n",
    "stat_res.summary()\n",
    "# Store the results in this dataframe\n",
    "dea_results = stat_res.results_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we visulize these results. First we plot a volcano plot higlighting the genes of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genes of Interest\n",
    "gene_list=[\"ZC3HAV1\",'AICDA', 'APOBEC1', 'APOBEC2', 'APOBEC3A', 'APOBEC3B', 'APOBEC3C', 'APOBEC3D', 'APOBEC3F', 'APOBEC3G', 'APOBEC3H', 'APOBEC4','ADAR', 'ADARB1', 'ADARB2']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plot_volcano(data, filtered_gene_list, logfc_threshold=1, padj_threshold=0.1, adjust=True):\n",
    "    \"\"\"\n",
    "    Plot a volcano plot to visualize differential gene expression analysis results.\n",
    "\n",
    "    Parameters:\n",
    "    - data (pandas.DataFrame): The data containing gene expression analysis results.\n",
    "    - filtered_gene_list (list): A list of genes of interest to be highlighted in the plot.\n",
    "    - logfc_threshold (float, optional): The threshold for log2 fold change. Default is 1.\n",
    "    - padj_threshold (float, optional): The threshold for adjusted p-value. Default is 0.1.\n",
    "    - adjust (bool, optional): Whether to adjust the position of gene labels to avoid overlap. Default is True.\n",
    "\n",
    "    \"\"\"\n",
    "    plt.figure(figsize=(10, 10))\n",
    "\n",
    "    # Calculate thresholds\n",
    "    padj_log_threshold = -np.log10(padj_threshold)\n",
    "\n",
    "    # Filter points based on thresholds\n",
    "    nonsig_points = data[(data['padj'] > padj_threshold) | (abs(data['log2FoldChange']) < logfc_threshold)]\n",
    "    sig_points = data[(data['padj'] <= padj_threshold) & (abs(data['log2FoldChange']) >= logfc_threshold)]\n",
    "\n",
    "    # Plot non-significant points\n",
    "    plt.scatter(nonsig_points['log2FoldChange'], -np.log10(nonsig_points['padj']), c='#DCDCDC', edgecolor='#E0E0E0', label='Not Statistically Significant', marker='o')\n",
    "\n",
    "    # Plot significant points\n",
    "    plt.scatter(sig_points['log2FoldChange'], -np.log10(sig_points['padj']), c='#A9A9A9', edgecolor='#E0E0E0', label='Statistically Significant', marker='o')\n",
    "\n",
    "    # Plot genes in the filtered gene list with a different marker\n",
    "    filtered_points = data[data.index.isin(filtered_gene_list)]\n",
    "    plt.scatter(filtered_points['log2FoldChange'], -np.log10(filtered_points['padj']), c='b', edgecolor='none', label='Genes of Interest', marker='^')\n",
    "\n",
    "    # Draw horizontal line for padj threshold\n",
    "    plt.axhline(y=-np.log10(padj_threshold), color='#909090', linestyle='--', linewidth=1)\n",
    "\n",
    "    # Draw vertical lines for log2 fold change thresholds\n",
    "    plt.axvline(x=logfc_threshold, color='#909090', linestyle='--', linewidth=1)\n",
    "    plt.axvline(x=-logfc_threshold, color='#909090', linestyle='--', linewidth=1)\n",
    "\n",
    "    # Label points in the filtered gene list\n",
    "    texts = []\n",
    "    for gene in filtered_gene_list:\n",
    "        gene_data = data[data.index == gene]\n",
    "        if not gene_data.empty:\n",
    "            x = gene_data['log2FoldChange'].values[0]\n",
    "            y = -np.log10(gene_data['padj'].values[0])\n",
    "            txt = plt.text(x, y, gene, size=15, ha='center', va='center')  # Decreased font size\n",
    "            txt.set_path_effects([PathEffects.withStroke(linewidth=3, foreground='w')])\n",
    "            texts.append(txt)\n",
    "\n",
    "    plt.legend(fontsize=15)  # Increased font size of the legend\n",
    "    plt.xlabel(\"$log_{2}$ fold change\", size=15)\n",
    "    plt.ylabel(\"-$log_{10}$ $p_{adj}$-value\", size=15)\n",
    "\n",
    "    if adjust:\n",
    "        adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color='k', lw=0.5))\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n",
    "# Add the GeneSymbol column to the DataFrame\n",
    "dea_results[\"GeneSymbol\"]=dea_results.index\n",
    "# Call the function to plot the volcano plot\n",
    "plot_volcano(dea_results, gene_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we carry out Gense set enrichment for the Oxidative Damage Response geneset using GSEApy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def run_gene_set_enrichment_analysis(dea_results, oxidative_stress_gene_set):\n",
    "    \"\"\"\n",
    "    Runs gene set enrichment analysis using GSEApyfor a specified gene set.\n",
    "\n",
    "    Args:\n",
    "        dea_results(pandas dataframe): DESeq2 results with columns 'GeneSymbol', 'padj', and 'log2FoldChange'.\n",
    "        oxidative_stress_gene_set (str): Path to the Oxidative Damage response gene set file.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a ranking of genes based on the DESeq2 results\n",
    "    dea_results[\"-log_padj x lfc\"] = -np.log(dea_results[\"padj\"]) * dea_results[\"log2FoldChange\"]\n",
    "    # Sort the genes by the -log_padj x lfc\n",
    "    ranked_genes = dea_results[['GeneSymbol', '-log_padj x lfc']].dropna().sort_values('-log_padj x lfc', ascending=False)\n",
    "    # Reset the index to GeneSymbol column\n",
    "    ranked_genes.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    # Run Gene Set Enrichment Analysis\n",
    "    pre_res = gp.prerank(rnk=ranked_genes, gene_sets=oxidative_stress_gene_set, seed=6, min_size=0, max_size=200)\n",
    "\n",
    "    # Plot the results\n",
    "    pre_res.plot(terms=\"WP_OXIDATIVE_DAMAGE_RESPONSE\", figsize=(10, 10))\n",
    "\n",
    "\n",
    "run_gene_set_enrichment_analysis(dea_results, \"WP_OXIDATIVE_DAMAGE_RESPONSE.v2023.2.Hs.gmt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to invesgate the variability of the genes of interest among different replicates we plot the clustermap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plot_heatmap_with_clustermap(dds, dea_results, gene_list, metadata):\n",
    "    \"\"\"\n",
    "    Plot a heatmap using the genes of interest for the different replicates.\n",
    "\n",
    "    Args:\n",
    "        dds (object): The DESeq2 object containing normalized counts and log1p transformed data.\n",
    "        dea_results (DataFrame): The Dataframe contatining the DESeq2 results.\n",
    "        gene_list (list): The list of genes to be used in constructing the heatmap.\n",
    "        metadata (DataFrame): The DataFrame containing metadata information of the replicates.\n",
    "\n",
    "    \"\"\"\n",
    "    # Extract the normalized counts and log1p transform them with a pseudocount of 1\n",
    "    dds.layers['log1p'] = np.log1p(dds.layers['normed_counts'])\n",
    "    dds_sigs = dds[:, dea_results[dea_results[\"GeneSymbol\"].isin(gene_list)].index]\n",
    "\n",
    "    # Create a DataFrame for plotting the heatmap\n",
    "    grapher = pd.DataFrame(dds_sigs.layers['log1p'].T, index=dds_sigs.var_names, columns=dds_sigs.obs_names)\n",
    "\n",
    "\n",
    "    # Get the indexes of control and covid conditions\n",
    "    indexes_condition_control = metadata[metadata['Condition'] == \"Control\"].index.tolist()\n",
    "    indexes_condition_covid = metadata[metadata['Condition'] == \"Covid\"].index.tolist()\n",
    "    indexes_order = indexes_condition_control + indexes_condition_covid\n",
    "\n",
    "    # Clean the data and reorder it based on gene_list and indexes_order\n",
    "    cleaned_data = grapher[indexes_order]\n",
    "    cleaned_data = cleaned_data.reindex(gene_list)\n",
    "    cleaned_data.fillna(0, inplace=True)\n",
    "    cleaned_data = cleaned_data[sorted(cleaned_data.columns)]\n",
    "\n",
    "    \n",
    "    # Create a DataFrame to use as col_colors\n",
    "    col_colors_df = pd.DataFrame(index=cleaned_data.columns)\n",
    "\n",
    "    # highlight the control and the covid samples with different colors\n",
    "    col_colors_df.loc[indexes_condition_control, 'color'] = 'blue'  # Color for control columns\n",
    "    col_colors_df.loc[indexes_condition_covid, 'color'] = 'red'  # Color for covid columns\n",
    "\n",
    "\n",
    "    # Generate clustermap with marked columns\n",
    "    cluster = sns.clustermap(cleaned_data, method='average', z_score=-1, cmap='RdYlBu_r', col_cluster=False, row_cluster=False, col_colors=[col_colors_df['color']],  figsize=(25, 10))\n",
    "    cluster.ax_heatmap.set_xticklabels(cluster.ax_heatmap.get_xticklabels(), fontsize=12)\n",
    "\n",
    "    # Create the legend \n",
    "    legend = [Patch(color=color, label=label) for label, color in {'Control': 'blue', 'COVID-19': 'red'}.items()]\n",
    "\n",
    "    # Add legend to the plot\n",
    "    plt.legend(handles=legend, loc='upper left', bbox_to_anchor=(0, -0.1), frameon=False)\n",
    "\n",
    "    # Show the plot\n",
    "    plt.show()\n",
    "    return cleaned_data\n",
    "\n",
    "# Call the function to plot the heatmap\n",
    "plot_heatmap_with_clustermap(dds, dea_results, gene_list, metadata)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At last we plot a heatmap depicting the log2Fold changes for the genes of interest "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract the LogFolChanges for the genes of interest\n",
    "log2_fold_changes = [dea_results[dea_results.index == gene][\"log2FoldChange\"].values[0] for gene in gene_list]\n",
    "log2_fold_changes_df=pd.DataFrame(data=log2_fold_changes, index=gene_list, columns=[\"log2FoldChange\"])\n",
    "\n",
    "# Plotting the heatmap\n",
    "# Set the figure size\n",
    "plt.figure(figsize=(.6, 10))\n",
    "# Create the heatmap  \n",
    "heatmap = sns.heatmap(log2_fold_changes_df, cmap='coolwarm', annot=True, fmt=\".2f\", cbar=False, annot_kws={\"size\": 9})\n",
    "# Set the title\n",
    "plt.title('Log$_2$ Fold Changes', fontweight='bold', fontsize=10) \n",
    "# Turn off x-axis ticks\n",
    "plt.xticks([])  \n",
    "# Set the font size for y-axis ticks\n",
    "heatmap.yaxis.set_tick_params(labelsize=10)  \n",
    "plt.tight_layout()  \n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the processed count matrix and metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the count matrix\n",
    "counts_df.to_csv('bulk_transcriptomic_analysis_processed_count_matrix.tsv', sep='\\t')\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mew",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
