{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import AlignIO\n",
    "from Bio import Phylo\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Seq import Seq\n",
    "from adjustText import adjust_text\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "from itertools import product\n",
    "import re\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def root_tree_with_first_sequence(tree, root_sequence):\n",
    "    \"\"\"\n",
    "    Roots the tree file with the provided root sequence.\n",
    "\n",
    "    Args:\n",
    "    tree: The tree object to be rooted.\n",
    "    root_sequence (str): The sequence_id to be used as the root.\n",
    "\n",
    "    Returns:\n",
    "    rooted_tree_file (str): The path to the rooted tree file.\n",
    "    \"\"\"\n",
    "    # Root the tree file with the outgroup\n",
    "    rooted_tree_file = tree.root_with_outgroup(root_sequence)\n",
    "\n",
    "    return rooted_tree_file\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reconstructed_ancestral_sequences(ancestral_state_file):\n",
    "    \"\"\"\n",
    "    Reconstructs ancestral sequences from the ancestral state file.\n",
    "    \n",
    "    Parameters:\n",
    "        ancestral_state_file (str): The path to the ancestral state file.\n",
    "    \n",
    "    Returns:\n",
    "        dict: A dictionary containing the reconstructed ancestral sequences, where the keys are the node names and the values are the sequences.\n",
    "    \"\"\"\n",
    "    # Read the ancestral state file\n",
    "    ancestral_states = pd.read_csv(ancestral_state_file, sep='\\t', skiprows=8)\n",
    "    # Concatenate the ancestral states at each position in a given node to reconstruct ancestral sequences\n",
    "    ancestral_sequences = {}\n",
    "    for node, group in ancestral_states.groupby('Node'):\n",
    "        sequence = ''.join(group['State'])\n",
    "        ancestral_sequences[node] = sequence\n",
    "    \n",
    "    return ancestral_sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to compute the trinucleotide frequency of a sequence to be used in normalization step while computing the trinucleotide mutation rate\n",
    "def get_kmer_frequency(sequence, k):\n",
    "    \"\"\"\n",
    "    Calculate the frequency of k-mers in a given sequence.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): The input RNA sequence.\n",
    "        k (int): The length of the k-mer.\n",
    "\n",
    "    Returns:\n",
    "        kmers_frequency_dict: A dictionary containing all possible k-mers as keys and their frequencies in the input sequence as values.\n",
    "    \"\"\"\n",
    "    # Remove gaps from the sequence\n",
    "    sequence = sequence.replace(\"-\", \"\")\n",
    "    #Total number of kmers in the sequence\n",
    "    total_kmers = len(sequence) - k + 1\n",
    "\n",
    "    # Declare and Initialize the dictionary to store the kmer frequency by assigning all possible k-mers with frequency 0\n",
    "    nucleotides = [\"C\", \"T\", \"G\", \"A\"]\n",
    "    kmers_frequency_dict = {''.join(combo): 0 for combo in product(nucleotides, repeat=k)}\n",
    "\n",
    "    # Iterate through the sequence to find all k-mers\n",
    "    for i in range(total_kmers):\n",
    "        kmer = sequence[i:i + k]  # Extract the k-mer\n",
    "        # Check if the k-mer consists only of A, T, G, or C and not any other characters\n",
    "        if set(kmer).issubset(set(nucleotides)):\n",
    "            # Increment the frequency of this k-mer\n",
    "            kmers_frequency_dict[kmer] += 1  \n",
    "\n",
    "    return kmers_frequency_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_substitution_mutations(ancestor_seq, sample_seq):\n",
    "    \"\"\"\n",
    "    To compute the mutations by comparing each sequence against its nearest reconstructed ancestral sequence\n",
    "\n",
    "    Args:\n",
    "        ancestor_seq (str): The ancestor sequence.\n",
    "        sample_seq (str): The sample sequence.\n",
    "\n",
    "    Returns:\n",
    "        mutations: A list of mutations, where each mutation is represented as a list containing the mutation, the trinucleotide context, and the trinucleotide normalized mutation rate.\n",
    "    \"\"\"\n",
    "    # Initialize the list to store the mutations\n",
    "    mutations = []\n",
    "    # Get the trinucleotide mutation frequencies in the ancestor sequence\n",
    "    trinucleotide_mutation_frequency = get_kmer_frequency(ancestor_seq, 3)\n",
    "    # Iterate through each position and base in the two sequences\n",
    "    for i, (a, b) in enumerate(zip(ancestor_seq, sample_seq)):\n",
    "        # If the base is different between the ancestor and sample sequences \n",
    "        if a != b:\n",
    "            #position of the mutation in the alignment is index + 1\n",
    "            position = i + 1  \n",
    "            # Extract the trinucleotide context of the mutation\n",
    "            trinucleotide = str(ancestor_seq[position-2:position+1]).upper()  \n",
    "            # Base in the ancestor sequence\n",
    "            ref_base = a.upper()\n",
    "            # Base in the sample sequence\n",
    "            sample_base = b.upper()\n",
    "            # We ignore those mutations at the starting or end poisition of the alignment as their trinucleotide context cannot be captured\n",
    "            # We also ignore those mutations in which the surrounding trinucleotide contains abnormal basses or the mutation itself contains abnormal bases\n",
    "            if (\n",
    "                any(base not in [\"C\", \"T\", \"G\", \"A\"] for base in trinucleotide) or\n",
    "                any(base not in [\"C\", \"T\", \"G\", \"A\"] for base in [ref_base, sample_base]) or\n",
    "                len(trinucleotide) != 3\n",
    "            ):                \n",
    "                continue\n",
    "            else:\n",
    "                # Record the mutation data which includes the mutation, its trinucleotide context and trinuleotide mutation rate (normalized) \n",
    "                mutation = f\"{ref_base}{position}{sample_base}\".upper()\n",
    "                mutations.append([mutation, trinucleotide, 1/trinucleotide_mutation_frequency[trinucleotide]])\n",
    "               \n",
    "    return mutations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_ancestral_node(tree, node_name):\n",
    "    \"\"\"\n",
    "    To find the nearest ancestor to a particular node in the tree\n",
    "\n",
    "    Args:\n",
    "        tree: The phylogenetic tree to search in.\n",
    "        node_name: The name of the node for which we wish to find the neareast ancestor.\n",
    "\n",
    "    Returns:\n",
    "        ancestral_node: The ancestral node if found, None otherwise.\n",
    "    \"\"\"\n",
    "    # Search the target node in the tree\n",
    "    target_node = next((clade for clade in tree.find_clades() if str(clade.name).split(\"/\")[0] == node_name), None)\n",
    "    if target_node is None:\n",
    "        print(f\"Node '{node_name}' not found.\")\n",
    "        return None\n",
    "\n",
    "    # Traverse the tree to find the ancestral node to the target node\n",
    "    ancestral_node = None\n",
    "    for clade in tree.find_clades():\n",
    "        if target_node in clade:\n",
    "            ancestral_node = clade\n",
    "            break\n",
    "\n",
    "    return ancestral_node\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_mutation_data_for_internal_nodes(ancestral_sequences, alignment_file, tree):\n",
    "    \"\"\" \n",
    "    To compute the mutations for the internal nodes of the tree\n",
    "\n",
    "    Args:\n",
    "        ancestral_sequences (dict): Dictionary of reconstructed ancestral sequences.\n",
    "        alignment_file (str): The path to the mutiple sequence alignment file.\n",
    "        tree: The phylogenetic tree.\n",
    "\n",
    "    Returns:\n",
    "        - denovo_mutations: A dictionary mapping each internal node to its denovo mutations.\n",
    "         - ancestral_mutations: A dictionary mapping each internal node to its ancestral mutations.\n",
    "         - ancestral_node_data: A dictionary mapping each internal node to its nearest ancestral node.\n",
    "\n",
    "    \"\"\"\n",
    "    # Initialize the dictionaries to store the mutation data\n",
    "    denovo_mutations = {}\n",
    "    ancestral_mutations = {}\n",
    "    ancestral_node_data = {}\n",
    "    # Sort the ancestral nodes based on their node number\n",
    "    ancestral_nodes = list(ancestral_sequences.keys())\n",
    "    ancestral_nodes = sorted(ancestral_nodes, key=lambda node: int(re.search(r'\\d+', node).group()))\n",
    "\n",
    "    for node in ancestral_nodes:\n",
    "        #If it is the initial node, then the reference sequence is used as the ancestral sequence\n",
    "        if node == \"Node1\":\n",
    "            reference_record = next(SeqIO.parse(alignment_file, \"fasta\"), None)\n",
    "            reference_id = reference_record.id\n",
    "            reference_seq = str(reference_record.seq).upper()\n",
    "            denovo_mutations[node] = find_substitution_mutations(reference_seq, ancestral_sequences[node])\n",
    "            ancestral_mutations[node] = []\n",
    "            ancestral_node_data[\"Node1\"] = reference_id\n",
    "        #If it is not the initial node, then we find the nearest ancestor and compute the mutations against it\n",
    "        else:\n",
    "            ancestral_node = find_ancestral_node(tree, node)\n",
    "            denovo_mutations[node] = find_substitution_mutations(\n",
    "                ancestral_sequences[str(ancestral_node).split(\"/\")[0]], ancestral_sequences[node]\n",
    "            )\n",
    "            ancestral_mutations[node] = (\n",
    "                denovo_mutations[str(ancestral_node).split(\"/\")[0]]\n",
    "                + ancestral_mutations[str(ancestral_node).split(\"/\")[0]]\n",
    "            )\n",
    "            ancestral_node_data[node] = ancestral_node\n",
    "\n",
    "    return denovo_mutations, ancestral_mutations, ancestral_node_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_mutation_data_for_leaf_nodes(alignment_file, ancestral_sequences, denovo_mutations, ancestral_mutations, tree, ancestral_node_data):\n",
    "    \"\"\"\n",
    "    To compute the mutations for the leaf nodes of the tree\n",
    "    \n",
    "    Parameters:\n",
    "    - alignment_file (str): Path to the multiple sequence alignment file .\n",
    "    - ancestral_sequences (dict): Dictionary of reconstructed ancestral sequences.\n",
    "    - denovo_mutations (dict): Dictionary to store denovo mutations. \n",
    "    - ancestral_mutations (dict): Dictionary to store ancestral mutations. \n",
    "    - tree: the Phylogenetic tree object.\n",
    "    - ancestral_node_data (dict): Dictionary to store ancestral node data.\n",
    "\n",
    "    Returns:\n",
    "    - denovo_mutations (dict): Updated dictionary of denovo mutations.\n",
    "    - ancestral_mutations (dict): Updated dictionary of ancestral mutations.\n",
    "    - ancestral_node_data (dict): Updated dictionary of ancestral node data for each leaf node.\n",
    "    \"\"\"\n",
    "    # Get the leaf node names from the alignment file    \n",
    "    leaf_nodes = [record.id for record in SeqIO.parse(alignment_file, \"fasta\")]\n",
    "    #As the first leaf node is the reference sequence, we need to handle it separately\n",
    "    denovo_mutations[leaf_nodes[0]] = []\n",
    "    ancestral_mutations[leaf_nodes[0]] = []\n",
    "    #Remove the reference sequence from the leaf nodes\n",
    "    leaf_nodes = leaf_nodes[1:]\n",
    "    #Compute the mutations for each leaf node\n",
    "    for leaf_node in leaf_nodes:\n",
    "        #find its nearest ancestral node data\n",
    "        ancestral_node = find_ancestral_node(tree, leaf_node)\n",
    "        #store the ancestral node data for the leaf node\n",
    "        ancestral_node_data[leaf_node] = ancestral_node\n",
    "        #extract the sequence of the leaf node\n",
    "        sequence = next((str(record.seq).upper() for record in SeqIO.parse(alignment_file, \"fasta\") if record.id == leaf_node), None)\n",
    "        #compute the mutations against the nearest ancestral sequence\n",
    "        denovo_mutations[leaf_node] = find_substitution_mutations(\n",
    "            ancestral_sequences[str(ancestral_node).split(\"/\")[0]], \n",
    "            sequence)\n",
    "        #bootstrapping during tree construction can modify the names of the nodes in the tree by including the bootstrap results, hence we need to extract only the node name from the tree\n",
    "        #and use it to save mutation data for that leafnode\n",
    "        ancestral_mutations[leaf_node] = denovo_mutations[str(ancestral_node).split(\"/\")[0]] + ancestral_mutations[str(ancestral_node).split(\"/\")[0]]\n",
    "    return denovo_mutations, ancestral_mutations, ancestral_node_data\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We access the precomputed MSA file and the precomputed tree file. As IQTREE does not automatically root the tree with the outgroup hence we do it explicitly. Also, we use the ancestral states at each ancestral nodes of the tree to reconstruct the probable ancestral sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#path to the MSA file\n",
    "alignment_file = 'SARS-CoV-2_18000+1_msa.fasta'\n",
    "\n",
    "#we load the phylognetic tree \n",
    "tree = Phylo.read(f\"Fast IQTREE/fast.treefile\", \"newick\")\n",
    "\n",
    "#we now root it by assigning the outgroup/reference sequence as the root of the tree\n",
    "outgroup_id = \"NC_045512.2\"  \n",
    "root_tree_with_first_sequence(tree, outgroup_id)\n",
    "\n",
    "#we now reconstruct the ancestral sequences of the tree from the ancestral state file\n",
    "ancestral_sequences=get_reconstructed_ancestral_sequences(f\"Fast IQTREE/fast.state\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we compute the mutations by comparing each sequence to its nearest reconstructed ancestor. At first, we compute the mutations for the internal nodes and then for the leaf nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we now compute the mutations for the internal nodes of the tree\n",
    "denovo_mutations, ancestral_mutations,ancestral_node_data=get_mutation_data_for_internal_nodes(ancestral_sequences,alignment_file, tree)\n",
    "\n",
    "#we now compute the mutations for the leaf nodes of the tree   \n",
    "denovo_mutations, ancestral_mutations,ancestral_node_data=get_mutation_data_for_leaf_nodes(alignment_file,ancestral_sequences, denovo_mutations, ancestral_mutations,tree,ancestral_node_data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we are only concerend about the mutations in the leaf nodes which represents the sampled genomes. Hence we filter out the data for the internal nodes and the outgroup genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we extract the mutations for the leaf nodes(sampled genomes) of the tree\n",
    "denovo_mutations_in_sampled_genomes = {key: value for key, value in denovo_mutations.items() if not key.startswith(\"Node\")}\n",
    "denovo_mutations_in_sampled_genomes.pop(outgroup_id)\n",
    "ancestral_mutations_in_sampled_genomes = {key: value for key, value in denovo_mutations.items() if not key.startswith(\"Node\")}\n",
    "ancestral_mutations_in_sampled_genomes.pop(outgroup_id)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We save these dictionaries for later use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open('denovo_mutations_in_sampled_genomes.pkl', 'wb') as f:\n",
    "    pickle.dump(denovo_mutations_in_sampled_genomes, f)\n",
    "\n",
    "with open(\"ancestral_mutations_in_sampled_genomes.pkl\", 'wb') as f:\n",
    "    pickle.dump(ancestral_mutations_in_sampled_genomes, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we use the de novo mutations to compute the overall mutation profile of the 18000 SARS-CoV-2 sampled genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_mutation_stats_table(mutation_data):\n",
    "    \"\"\"\n",
    "    Calculates the mutation count and average trinucleotide mutation rate for each mutation type.\n",
    "\n",
    "    Args:\n",
    "        mutation_data (dict): the dictionary containing de novo mutation data for each sampled genome.\n",
    "\n",
    "    Returns:\n",
    "        overall_mutation_df: A DataFrame containing the mutation count and average trinucleotide mutation rate for each mutation type.\n",
    "\n",
    "    \"\"\"\n",
    "    # Initialize the dictionary to store the mutation count and average trinucleotide mutation rate for each mutation type\n",
    "    nucleotides=[\"C\", \"T\", \"G\", \"A\"]\n",
    "    mutation_count_dict={}\n",
    "    for nucleotide1 in nucleotides:\n",
    "        for nucleotide2 in nucleotides:\n",
    "            if(nucleotide1!=nucleotide2):\n",
    "                mutation_count_dict[nucleotide1+\"->\"+nucleotide2]=[0,0]\n",
    "\n",
    "    # Compute the mutation count and average trinucleotide mutation rate for each mutation type\n",
    "    total_genomes=18000\n",
    "    for leaf_node in mutation_data:\n",
    "        for mutation in mutation_data[leaf_node]:\n",
    "            mutation_type=mutation[0][0]+\"->\"+mutation[0][-1]\n",
    "            mutation_count_dict[mutation_type][0]+=1\n",
    "            mutation_count_dict[mutation_type][1]+=mutation[2]*(10000/(total_genomes*16))\n",
    "\n",
    "\n",
    "    # Creating the DataFrame\n",
    "    overall_mutation_df = pd.DataFrame(mutation_count_dict, index=['Mutation Count',  'Mean Trinucleotide Mutation Rate']).T\n",
    "\n",
    "    # Renaming the columns\n",
    "    overall_mutation_df.index.name = 'Mutation Type'\n",
    "\n",
    "    # Replacing the T with U in the index for the Mutation Type because we are dealing with RNA virus\n",
    "    overall_mutation_df.index = overall_mutation_df.index.str.replace('T', 'U')\n",
    "    \n",
    "    # Sorting the DataFrame based on the Mutation Count\n",
    "    overall_mutation_df = overall_mutation_df.sort_values(by='Mutation Count', ascending=False)\n",
    "    \n",
    "    \n",
    "    return overall_mutation_df\n",
    "\n",
    "overall_mutation_df=get_mutation_stats_table(denovo_mutations_in_sampled_genomes)\n",
    "\n",
    "print(\"Total mutation count:\", overall_mutation_df['Mutation Count'].sum())\n",
    "overall_mutation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_trinucleotide_mutation(denovo_mutations):\n",
    "    \"\"\"\n",
    "    Plots the trinucleotide mutation Rate for each mutation type.\n",
    "\n",
    "    Args:\n",
    "        denovo_mutations (dict): The dictionary containing the denovo mutations data for the sampled genomes.\n",
    "\n",
    "    \"\"\"\n",
    "    # Generate all possible trinucleotide combinations\n",
    "    flank_length=1\n",
    "    nucleotides=[\"C\", \"T\", \"G\", \"A\"]\n",
    "    trinucleotide_list=nucleotides.copy()\n",
    "    for i in range(flank_length*2):\n",
    "        new_list=[]\n",
    "        for base in nucleotides:\n",
    "            for old_index in trinucleotide_list.copy():\n",
    "                new_index=old_index[:int(len(old_index)/2)]+base+old_index[int(len(old_index)/2):]\n",
    "                new_list.append(new_index)\n",
    "        trinucleotide_list=new_list        \n",
    "        \n",
    "    # Initialize the dictionary to store the trinucleotide mutation rates for each mutation type\n",
    "    combined_mutation_data={}\n",
    "    for index in trinucleotide_list:\n",
    "        combined_mutation_data[index]={}\n",
    "        for nucleotide in nucleotides:\n",
    "            if(index[flank_length]!=nucleotide):\n",
    "                combined_mutation_data[index][nucleotide]=[]\n",
    "                \n",
    "    #Compute the trinucleotide mutation rate for each mutation type\n",
    "    for genome in denovo_mutations:\n",
    "        for mutation_in_genome in denovo_mutations[genome]:\n",
    "            trinucleotide=mutation_in_genome[1]\n",
    "            alternate_base=mutation_in_genome[0][-1]\n",
    "            combined_mutation_data[trinucleotide][alternate_base].append(mutation_in_genome)\n",
    "            \n",
    "    # Generate the mutation signature list from the combined mutation data to help in plotting\n",
    "    mutation_signature_list=[]\n",
    "    #subheadings to be displayed at the top of every 16 bars(representing the trinucleotide mutation rates) are the mutation types associated these trinucleotide muation rates\n",
    "    subheadings=[]\n",
    "    #Iterating through the combined mutation data to generate the mutation signature list\n",
    "    for i in range(0,len(combined_mutation_data),4**(flank_length*2)):\n",
    "            sublist=list(combined_mutation_data.keys())[i:i+4**(flank_length*2)]\n",
    "            if(flank_length!=0):\n",
    "                if(sublist[flank_length][1] in nucleotides):\n",
    "                    for new_base in [\"A\",\"C\",\"G\",\"T\"]:\n",
    "                        if(sublist[0][flank_length]!=new_base):\n",
    "                            subheadings.append((sublist[0][flank_length]+\"->\"+new_base).replace(\"T\",\"U\"))\n",
    "                        for polynucleotide in sublist:\n",
    "                            if(polynucleotide[flank_length]!=new_base):\n",
    "                                mutation_signature_list.append([polynucleotide, new_base,10000*sum(i[2] for i in combined_mutation_data[polynucleotide][new_base])/(len(denovo_mutations)),len(combined_mutation_data[polynucleotide][new_base])])\n",
    "            else:\n",
    "                if(sublist[flank_length] in nucleotides):\n",
    "                    for new_base in [\"A\",\"C\",\"G\",\"T\"]:\n",
    "                        if(sublist[flank_length]!=new_base):\n",
    "                            subheadings.append((sublist[0][flank_length]+\"->\"+new_base).replace(\"T\",\"U\"))\n",
    "                        for polynucleotide in sublist:\n",
    "                            if(polynucleotide[flank_length]!=new_base):\n",
    "                                mutation_signature_list.append([polynucleotide, new_base, len(combined_mutation_data[polynucleotide][new_base])])\n",
    "                \n",
    "\n",
    "\n",
    "    # Extracting mutation types and counts\n",
    "    mutation_types = [mutation[0].replace(\"T\",\"U\")+\"->\"+mutation[1].replace(\"T\",\"U\") for mutation in mutation_signature_list] # Replace 'T' with 'U' for RNA\n",
    "    mutation_counts = [mutation[2] for mutation in mutation_signature_list]\n",
    "    mutation_count_fraction=mutation_counts\n",
    "    print(mutation_types)\n",
    "    print(mutation_counts)\n",
    "\n",
    "    # Define colors for every 16 bars\n",
    "    colors = ['#5fbceb', '#06070a', '#d33c33', '#cacaca', '#accc6f', '#e7cac4', '#a5d4e6', '#5e6164', '#cd7430', '#eae8eb', '#dff488', '#f1dcdb']\n",
    "    color_cycle = [color for color in colors for _ in range(16)]\n",
    "\n",
    "\n",
    "    # Plotting the bar plot with cycling colors and gaps\n",
    "    plt.figure(figsize=(30, 7))  \n",
    "    for i in range(len(mutation_types)):\n",
    "            plt.bar(mutation_types[i], mutation_counts[i], color=color_cycle[i], edgecolor='none')  # Set edgecolor to 'none'\n",
    "\n",
    "            # Add heading at the top middle of every 16 bars\n",
    "            if (i + 1) % 16 == 8:  # Assuming the middle point is the 8th bar in each group\n",
    "                bbox_props = dict(boxstyle=\"round,pad=0.5\", edgecolor=color_cycle[i], facecolor=color_cycle[i], alpha=0.5)\n",
    "                plt.text(i, max(mutation_counts)+1, subheadings[int((i+1)/16)],\n",
    "                        ha='center', va='bottom', fontsize=12, color='black', bbox=bbox_props)\n",
    "            \n",
    "            # Add value(i.e. trinucleotide mutation rates) at the top of each bar\n",
    "            plt.text(i, mutation_counts[i], \"  \"+str(round(mutation_count_fraction[i],2)), ha='center', va='bottom', fontsize=10, rotation=90)\n",
    "\n",
    "    plt.xlabel('Trinuclotides', fontsize=12)\n",
    "    plt.ylabel(r'Average Trinucleotide Mutation Rate (Normalized) ($10^{-4}$) $\\rightarrow$', fontsize=12)\n",
    "    # Rotate x-axis labels for better visibility\n",
    "    plt.xticks(rotation=90, ha='center', fontsize=10)  \n",
    "    # Remove outline at top and right\n",
    "    plt.gca().spines['top'].set_visible(False)\n",
    "    plt.gca().spines['right'].set_visible(False)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Display the plot\n",
    "    plt.show()\n",
    "\n",
    "plot_trinucleotide_mutation(denovo_mutations_in_sampled_genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import logomaker\n",
    "\n",
    "def plot_trinucleotide_mutations_per_strand_bar_plot(mutation_data, reference_mutation_type, color1, complementary_mutation_type, color2):\n",
    "    \"\"\" \n",
    "    In this function we compare two complemantry muttions types based on the trinucleotide mutation rates\n",
    "    We plot the trinucleotide mutation rates for the reference and complementary mutation types in a bar plot both in sorted and unsorted manner\n",
    "    We also plot the Position Weight Matrix(PWM) computed from the trinucleotide mutation rate for the reference and complementary mutation types\n",
    "\n",
    "    Args: \n",
    "    - mutation_data (dict): The dictionary containing the denovo mutations data for the sampled genomes.\n",
    "    - reference_mutation_type (str): The reference mutation type.\n",
    "    - color1 (str): The color for representing the reference_mutation_type.\n",
    "    - complementary_mutation_type (str): The complementary mutation type.\n",
    "    - color2 (str): The color for representing the complementary_mutation_type.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Generate all possible trinucleotide combinations\n",
    "    nucleotides=[\"C\", \"T\", \"G\", \"A\"] \n",
    "    trinucleotide_list=nucleotides.copy()\n",
    "    for i in range(1*2):\n",
    "        new_list=[]\n",
    "        for base in nucleotides:\n",
    "            for old_index in trinucleotide_list.copy():\n",
    "                new_index=old_index[:int(len(old_index)/2)]+base+old_index[int(len(old_index)/2):]\n",
    "                new_list.append(new_index)\n",
    "        trinucleotide_list=new_list        \n",
    "\n",
    "    # Initialize the dictionaries to store the trinucleotide mutation rates for the different mutation types\n",
    "    count_data={}\n",
    "    cumulative_count={}\n",
    "    for trinucleotide in trinucleotide_list:\n",
    "        count_data[trinucleotide]={}\n",
    "        cumulative_count[trinucleotide]={}\n",
    "        for nucleotide in nucleotides:\n",
    "            if(trinucleotide[1]!=nucleotide):\n",
    "                count_data[trinucleotide][nucleotide]=0\n",
    "                cumulative_count[trinucleotide][nucleotide]=[]\n",
    "\n",
    "    # Compute the trinucleotide mutation rates for each mutation type\n",
    "    for this_genome in mutation_data:\n",
    "        if(this_genome in mutation_data):\n",
    "            trinucleotide_mutations_in_this_genome=mutation_data[this_genome]\n",
    "            for trinucleotide_mutation in trinucleotide_mutations_in_this_genome:\n",
    "                trinucleotide=trinucleotide_mutation[1]\n",
    "                alternate_base=trinucleotide_mutation[0][-1]\n",
    "                trinucleotide_muttion_rate=trinucleotide_mutation[2]\n",
    "                count_data[trinucleotide][alternate_base]+=trinucleotide_muttion_rate/(len(mutation_data))\n",
    "\n",
    "\n",
    "\n",
    "    # WE PLOT 6 SUBPLOTS(2 ROWS AND 3 COLUMNS)\n",
    "\n",
    "    # Specify the width ratios for each column\n",
    "    width_ratios = [2, 1,1]  \n",
    "    # Create subplots with different column widths\n",
    "    fig, ax = plt.subplots(2, 3, figsize=(20, 10), gridspec_kw={'width_ratios': width_ratios})\n",
    "\n",
    "    #Compute the trinucleotide mutation rates for the reference and complementary mutation types\n",
    "    #for the reference mutation type\n",
    "    trinucleotide_mutation_rates={}\n",
    "    for trinucleotide in count_data:\n",
    "        if(trinucleotide[1]==reference_mutation_type[0]):   \n",
    "            for alternate_base in count_data[trinucleotide]:\n",
    "                if(alternate_base==reference_mutation_type[-1]):   \n",
    "                    trinucleotide_mutation_rates[trinucleotide]=[count_data[trinucleotide][alternate_base]*10000]\n",
    "    #for the complementary mutation type\n",
    "    for trinucleotide in count_data:\n",
    "        if(trinucleotide[1]==complementary_mutation_type[0]):   \n",
    "            for alternate_base in count_data[trinucleotide]:\n",
    "                if(alternate_base==complementary_mutation_type[-1]):   \n",
    "                    trinucleotide_mutation_rates[str(Seq(trinucleotide).reverse_complement())].append(count_data[trinucleotide][alternate_base]*10000)    \n",
    "\n",
    "\n",
    "\n",
    "    #PLOTTING THE TRINUCLEOTIDE MUTATATION RATES BAR PLOTS FOR THE REFERENCE AND COMPLEMENTARY MUTATION TYPES\n",
    "    # Plotting the bar plot for the reference mutation type\n",
    "    bars1 = ax[0,0].bar([f\"{polynucleotide}_{str(Seq(polynucleotide).reverse_complement())}\" for polynucleotide in trinucleotide_mutation_rates],\n",
    "                    [trinucleotide_mutation_rates[polynucleotide][0] for polynucleotide in trinucleotide_mutation_rates],\n",
    "                    color=color1)\n",
    "    ax[0,0].set_ylabel(r'Average Trinucleotide Mutation Rate (Normalized) ($10^{-4}$) $\\rightarrow$', fontsize=8)\n",
    "    # Set the title as the reference mutation type\n",
    "    ax[0,0].set_title(reference_mutation_type.replace(\"T\",\"U\"), fontweight='bold', fontsize=18) #Replace 'T' with 'U' for RNA\n",
    "    # Set y-axis limit to 1.5 times the maximum trinucleotide mutation rate\n",
    "    ax[0,0].set_ylim(0, 1.5*max(trinucleotide_mutation_rates[polynucleotide][0] for polynucleotide in trinucleotide_mutation_rates))  \n",
    "    # Set x-axis labels\n",
    "    ax[0,0].set_xticklabels([f\"{polynucleotide} ({str(Seq(polynucleotide).reverse_complement())})\".replace(\"T\",\"U\") for polynucleotide in trinucleotide_mutation_rates], rotation=270) #Replace 'T' with 'U' for RNA\n",
    "    # Plotting the bar plot for the complementary mutation type\n",
    "    bars2 = ax[1,0].bar([f\"{polynucleotide}_{str(Seq(polynucleotide).reverse_complement())}\" for polynucleotide in trinucleotide_mutation_rates],\n",
    "                    [trinucleotide_mutation_rates[polynucleotide][1] for polynucleotide in trinucleotide_mutation_rates],\n",
    "                    color=color2)\n",
    "    ax[1,0].set_ylabel(r'Average Trinucleotide Mutation Rate (Normalized) ($10^{-4}$) $\\rightarrow$', fontsize=8)\n",
    "    # Set the title as the complementary mutation type\n",
    "    ax[1,0].set_title(complementary_mutation_type.replace(\"T\",\"U\"), fontweight='bold', fontsize=18) #Replace 'T' with 'U' for RNA\n",
    "    # Set x-axis label\n",
    "    ax[1,0].set_xlabel('Trinucleotides ->')\n",
    "    # Set y-axis limit to 1.5 times the maximum trinucleotide mutation rate\n",
    "    ax[1,0].set_ylim(0, 1.5*max(trinucleotide_mutation_rates[polynucleotide][1] for polynucleotide in trinucleotide_mutation_rates))  \n",
    "    # Set x-axis labels\n",
    "    ax[1,0].set_xticklabels([f\"{str(Seq(polynucleotide).reverse_complement())} ({polynucleotide})\".replace(\"T\",\"U\") for polynucleotide in trinucleotide_mutation_rates], rotation=270) #Replace 'T' with 'U' for RNA\n",
    "    \n",
    "    # Add a horizontal line deicting the average trinucleotide mutation rate\n",
    "    ax[0,0].axhline(np.mean([bar.get_height() for bar in bars1]), color='black', linestyle='--', linewidth=1)\n",
    "    ax[1,0].axhline(np.mean([bar.get_height() for bar in bars2]), color='black', linestyle='--', linewidth=1)\n",
    "    # Function to mention the associated Trinucleotide Mutation Rate on top of each bar\n",
    "    def add_value_labels(ax, bars, counts):\n",
    "        bar_count = -1\n",
    "        for bar in bars:\n",
    "            bar_count += 1\n",
    "            height = bar.get_height()\n",
    "            ax.annotate('{}'.format(counts[bar_count]),\n",
    "                        xy=(bar.get_x() + bar.get_width() / 2, height),\n",
    "                        xytext=(0, 3), \n",
    "                        textcoords=\"offset points\",\n",
    "                        ha='center', va='bottom', rotation=90)  \n",
    "    # Add Trinucleotide Mutation rate value at the top of each bar in both plots\n",
    "    add_value_labels(ax[0,0], bars1, [round(trinucleotide_mutation_rates[polynucleotide][0],2) for polynucleotide in trinucleotide_mutation_rates])\n",
    "    add_value_labels(ax[1,0], bars2, [round(trinucleotide_mutation_rates[polynucleotide][1],2) for polynucleotide in trinucleotide_mutation_rates])\n",
    "\n",
    "\n",
    "\n",
    "    #WE PLOT THE POSITION WEIGHT MATRIX FOR THE REFERENCE MUTATION TYPE\n",
    "    combined_mutation_dict={}\n",
    "    for trinucleotide in count_data:\n",
    "        if(trinucleotide[1]==reference_mutation_type[0]):\n",
    "            for alternate_base in count_data[trinucleotide]:\n",
    "                if(alternate_base==reference_mutation_type[-1]):\n",
    "                    combined_mutation_dict[trinucleotide]=count_data[trinucleotide][alternate_base]\n",
    "    #Initialize the Position Weight Matrix(PWM) for the reference mutation type\n",
    "    pwm = np.zeros((3, 4))\n",
    "    #Compute the Position Weight Matrix(PWM) for the reference mutation type\n",
    "    for trinucleotide in combined_mutation_dict:\n",
    "        rate=combined_mutation_dict[trinucleotide]\n",
    "        if(rate>0):\n",
    "            for i in range(0, len(trinucleotide)):\n",
    "                if trinucleotide[i] == \"A\":\n",
    "                    pwm[i, 0] += rate\n",
    "                elif trinucleotide[i] == \"C\":\n",
    "                    pwm[i, 1] += rate\n",
    "                elif trinucleotide[i] == \"G\":\n",
    "                    pwm[i, 2] += rate\n",
    "                elif trinucleotide[i] == \"T\":\n",
    "                    pwm[i, 3] += rate\n",
    "    # Normalize the PWM base frquencies\n",
    "    normalized_pwm = pwm / pwm.sum(axis=1, keepdims=True)\n",
    "    # Create a DataFrame from the Normalized PWM \n",
    "    pwm_df = pd.DataFrame(normalized_pwm, index=[i for i in range(1, 4)], columns=[\"A\", \"C\", \"G\", \"U\"])\n",
    "\n",
    "\n",
    "    # Create a Logo object to Visualize the Position Weight Matrix(PWM)\n",
    "    logo_reference_strand = logomaker.Logo(pwm_df, ax=ax[0,2], shade_below=0.5, fade_below=0.5)\n",
    "    # Style for better visualization\n",
    "    logo_reference_strand.style_spines(visible=False)\n",
    "    logo_reference_strand.style_spines(spines=['left', 'bottom'], visible=True)\n",
    "    logo_reference_strand.style_xticks(rotation=90, fmt='%d', anchor=0)\n",
    "    logo_reference_strand.ax.xaxis.set_ticks_position('none')\n",
    "    logo_reference_strand.ax.xaxis.set_tick_params(pad=-1)\n",
    "    # add it to the plot\n",
    "    logo_reference_strand.draw()\n",
    "\n",
    "\n",
    "\n",
    "    #PLOT THE SORTED TRINUCLEOTIDE MUTATION RATES FOR THE REFERENCE MUTATION TYPE\n",
    "    # Sort the trinucleotide mutation rates in descending order\n",
    "    sorted_trinucleotide_mutation_rate_data = dict(sorted(combined_mutation_dict.items(), key=lambda item: item[1], reverse=True))\n",
    "    # Extract trinucelotide labels and corresponding mutation rates for plotting\n",
    "    trinucelotides = [f\"{item} ({str(Seq(item).reverse_complement())})\".replace(\"T\",\"U\") for item in sorted_trinucleotide_mutation_rate_data]\n",
    "    trinucelotide_mutation_rates = [sorted_trinucleotide_mutation_rate_data[item]*10000 for item in sorted_trinucleotide_mutation_rate_data]\n",
    "    # Plotting\n",
    "    ax[0, 1].barh(trinucelotides, trinucelotide_mutation_rates, color=color1)  \n",
    "    #Styling\n",
    "    ax[0, 1].set_title(reference_mutation_type.replace(\"T\",\"U\"), fontweight='bold', fontsize=18)\n",
    "    ax[0, 1].invert_yaxis()  # Invert y-axis for readability\n",
    "    ax[0, 1].tick_params(axis='x', rotation=90)  # Rotate x-ticks vertically\n",
    "    \n",
    "    \n",
    "    \n",
    "    #WE PLOT THE POSITION WEIGHT MATRIX FOR THE COMPLEMENTARY MUTATION TYPE\n",
    "    #Compute the trinucleotide mutation rates for the complementary mutation type\n",
    "    combined_mutation_dict={}\n",
    "    for trinucleotide in count_data:\n",
    "        if(trinucleotide[1]==complementary_mutation_type[0]):\n",
    "            for alternate_base in count_data[trinucleotide]:\n",
    "                if(alternate_base==complementary_mutation_type[-1]):\n",
    "                    combined_mutation_dict[trinucleotide]=count_data[trinucleotide][alternate_base]\n",
    "    # Initialize the Position Weight Matrix(PWM) for the complementary mutation type\n",
    "    pwm = np.zeros((3, 4))\n",
    "    # Compute the Position Weight Matrix(PWM) for the complementary mutation type based on trinucleotide mutation rates\n",
    "    for trinucleotide in combined_mutation_dict:\n",
    "        rate=combined_mutation_dict[trinucleotide]\n",
    "        if(rate>0):\n",
    "            for i in range(0, len(trinucleotide)):\n",
    "                if trinucleotide[i] == \"A\":\n",
    "                    pwm[i, 0] += rate\n",
    "                elif trinucleotide[i] == \"C\":\n",
    "                    pwm[i, 1] += rate\n",
    "                elif trinucleotide[i] == \"G\":\n",
    "                    pwm[i, 2] += rate\n",
    "                elif trinucleotide[i] == \"T\":\n",
    "                    pwm[i, 3] += rate\n",
    "    # Normalize the PWM base frquencies\n",
    "    normalized_pwm = pwm / pwm.sum(axis=1, keepdims=True)\n",
    "    # Create a DataFrame from the Normalized PWM \n",
    "    pwm_df = pd.DataFrame(normalized_pwm, index=[i for i in range(1, 4)], columns=[\"A\", \"C\", \"G\", \"U\"])\n",
    "    # Create a Logo object to Visualize the Position Weight Matrix(PWM)\n",
    "    logo_opposite_strand = logomaker.Logo(pwm_df, ax=ax[1,2], shade_below=0.5, fade_below=0.5)\n",
    "    # Style for better visualization\n",
    "    logo_opposite_strand.style_spines(visible=False)\n",
    "    logo_opposite_strand.style_spines(spines=['left', 'bottom'], visible=True)\n",
    "    logo_opposite_strand.style_xticks(rotation=90, fmt='%d', anchor=0)\n",
    "    logo_opposite_strand.ax.xaxis.set_ticks_position('none')\n",
    "    logo_opposite_strand.ax.xaxis.set_tick_params(pad=-1)\n",
    "\n",
    "\n",
    "\n",
    "    #PLOT THE SORTED TRINUCLEOTIDE MUTATION RATES FOR THE COMPLEMENTARY MUTATION TYPE\n",
    "    sorted_data = dict(sorted(combined_mutation_dict.items(), key=lambda item: item[1], reverse=True))\n",
    "    \n",
    "    # Extract trinucleotides and corresponding trinucleotide mutation rates for plotting\n",
    "    trinucleotides = [f\"{item} ({str(Seq(item).reverse_complement())})\".replace(\"T\",\"U\") for item in sorted_data]\n",
    "    trinucelotide_mutation_rates = [sorted_data[item]*10000 for item in sorted_data]\n",
    "\n",
    "    # Plotting\n",
    "    ax[1, 1].barh(trinucleotides, trinucelotide_mutation_rates, color=color2)  \n",
    "    ax[1, 1].set_xlabel(r'Average Trinucleotide Mutation Rate (Normalized) ($10^{-4}$) $\\rightarrow$')  \n",
    "    ax[1, 1].invert_yaxis()  # Invert y-axis for readability\n",
    "    ax[1, 1].tick_params(axis='x', rotation=90)  # Rotate x-ticks vertically\n",
    "    ax[1, 1].set_title(complementary_mutation_type.replace(\"T\", \"U\"), fontweight='bold', fontsize=18)\n",
    "\n",
    "\n",
    "    # Adjust the space between subplots\n",
    "    fig.subplots_adjust(hspace=0.45, wspace=0.3)  \n",
    "\n",
    "\n",
    "plot_trinucleotide_mutations_per_strand_bar_plot(denovo_mutations_in_sampled_genomes,  r\"C$\\rightarrow$T\", '#d33c33', r\"G$\\rightarrow$A\", '#a5d4e6')\n",
    "plot_trinucleotide_mutations_per_strand_bar_plot(denovo_mutations_in_sampled_genomes, r\"A$\\rightarrow$G\", '#dff488', r\"T$\\rightarrow$C\", '#accc6f')\n",
    "plot_trinucleotide_mutations_per_strand_bar_plot(denovo_mutations_in_sampled_genomes, r\"G$\\rightarrow$T\", '#cd7430', r\"C$\\rightarrow$A\", '#5fbceb')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mew",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
