{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import median_abs_deviation\n",
    "import pandas as pd\n",
    "import scvi\n",
    "import scanpy as sc\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import matplotlib.patheffects as PathEffects\n",
    "from adjustText import adjust_text\n",
    "import matplotlib.pyplot as plt\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "import seaborn as sns\n",
    "import celltypist\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the first step we access the raw read count matricies for each of the samples, carry out preprocessing to filter out low quality cells and them combine them into a single annotation data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outlier_removal(anndata):\n",
    "    \"\"\"\n",
    "    Perform quality control on the input AnnData object by filtering out outliers with median absolute deviation greater or less than 5.\n",
    "\n",
    "    Args:\n",
    "        anndata: The input annotation data matrix \n",
    "    Returns:\n",
    "        annData: The filtered Annotation data matrix after quality control.\n",
    "\n",
    "    \"\"\"\n",
    "    # annotate the mitochondrial genes\n",
    "    anndata.var[\"mt\"] = anndata.var_names.str.startswith(\"MT-\")\n",
    "    # annotate the ribosomal genes\n",
    "    anndata.var[\"ribo\"] = anndata.var_names.str.startswith((\"RPS\", \"RPL\"))\n",
    "    #computer quality control metrics\n",
    "    sc.pp.calculate_qc_metrics(\n",
    "        anndata, qc_vars=[\"mt\", \"ribo\"], inplace=True, percent_top=[20], log1p=True\n",
    "    )\n",
    "    #filter cells with median absolute deviation greater or less than 5\n",
    "    def is_outlier(anndata, metric: str, nmads: int):\n",
    "        M = anndata.obs[metric]\n",
    "        outlier = (M < np.median(M) - nmads * median_abs_deviation(M)) | (\n",
    "            np.median(M) + nmads * median_abs_deviation(M) < M\n",
    "        )\n",
    "        return outlier\n",
    "    \n",
    "    #annotate the outliers\n",
    "    anndata.obs[\"outlier\"] = (\n",
    "        is_outlier(anndata, \"log1p_total_counts\", 5)\n",
    "        | is_outlier(anndata, \"log1p_n_genes_by_counts\", 5)\n",
    "        | is_outlier(anndata, \"pct_counts_in_top_20_genes\", 5)\n",
    "    )\n",
    "    #filter out the outliers\n",
    "    anndata = anndata[(~anndata.obs.outlier)].copy()\n",
    "\n",
    "    return anndata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_combined_anndata(folder):\n",
    "    \"\"\"\n",
    "    Each of the raw count matrices from the indiviudal samples were accessed and preprocessed to remove doublets, outliers, and low quality cells. \n",
    "   \n",
    "    Parameters:\n",
    "    - folder (str): Folder path containing raw read count matrix files from the different samples.\n",
    "\n",
    "    Returns:\n",
    "    - anndata_list (list): A list of filtered Annotation Data objects generated from raw count matrices.\n",
    "    \"\"\"\n",
    "\n",
    "    anndata_list=[]\n",
    "    #Iterate through the raw count matrix files\n",
    "    for filename in os.listdir(folder):\n",
    "        #Print the filename\n",
    "        print(filename, filename.split(\"_\")[1][-3:])\n",
    "        #Access the .csv raw count matrices \n",
    "        if filename.endswith(\".csv\"):\n",
    "            #Load the data\n",
    "            anndata = sc.read_csv(folder + \"/\" + filename).T\n",
    "            #Print the initial number of cells\n",
    "            print(\"initial number of cells:\", len(anndata.obs))\n",
    "            #Remove duplicate genes\n",
    "            anndata.var_names_make_unique()\n",
    "            \n",
    "\n",
    "            #DOUBLET REMOVAL\n",
    "            #Filter out low quality cells\n",
    "            sc.pp.filter_genes(anndata, min_cells = 10)\n",
    "            sc.pp.filter_cells(anndata, min_genes=200)\n",
    "            #Subsetting the top 2000 highly variable genes\n",
    "            sc.pp.highly_variable_genes(anndata, flavor = 'seurat_v3'\n",
    "                                        ,n_top_genes = 2000, subset = True)\n",
    "            #Train scvi model\n",
    "            scvi.model.SCVI.setup_anndata(anndata)\n",
    "            vae = scvi.model.SCVI(anndata)\n",
    "            vae.train()\n",
    "            #Train SOLO model for doublet prediction\n",
    "            solo = scvi.external.SOLO.from_scvi_model(vae)\n",
    "            solo.train()\n",
    "            #Predict doublets\n",
    "            doublet_predition_results_df = solo.predict()\n",
    "            doublet_predition_results_df['prediction'] = solo.predict(soft = False)\n",
    "            #Select doublets with a difference of more than 1 between the doublet and singlet prediction scores\n",
    "            doublet_predition_results_df['dif'] = doublet_predition_results_df.doublet - doublet_predition_results_df.singlet\n",
    "            doublets = doublet_predition_results_df[(doublet_predition_results_df.prediction == 'doublet') & (doublet_predition_results_df.dif > 1)]\n",
    "            #Reload the data\n",
    "            anndata = sc.read_csv(folder + \"/\" + filename).T\n",
    "            anndata.obs_names_make_unique()\n",
    "            #Remove the doublets\n",
    "            anndata.obs['doublet'] = anndata.obs.index.isin(doublets.index)\n",
    "            anndata = anndata[~anndata.obs.doublet]\n",
    "            print(\"Cells after doublet removal:\", len(anndata.obs))\n",
    "\n",
    "\n",
    "            #ASSIGN SAMPLE NAME\n",
    "            #Extract Sample name from the filename\n",
    "            anndata.obs[\"Sample\"]=filename.split(\"_\")[1]\n",
    "\n",
    "            \n",
    "            #FILTER OUTLIERS\n",
    "            anndata=outlier_removal(anndata)\n",
    "            print(\"Cells after qc filtering removal:\", len(anndata.obs))\n",
    "            \n",
    "            #Append the filtered AnnData object to the list\n",
    "            anndata_list.append(anndata)\n",
    "\n",
    "            print(f\" successfully loaded {folder}/{filename}\")\n",
    "          \n",
    "    return anndata_list\n",
    "\n",
    "#Generate the combined AnnData object\n",
    "combined_anndata = sc.concat(generate_combined_anndata(\"../Covid-19-sc\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ADDITIONAL FILTERING TO REMOVE LOW QUALITY CELLS\n",
    "\n",
    "#Filter out cells with more than 10% mitochondrial genes\n",
    "combined_anndata = combined_anndata[combined_anndata.obs['pct_counts_mt'] <= 10]\n",
    "#Filter out cells with more than 10% ribosomal genes\n",
    "combined_anndata = combined_anndata[combined_anndata.obs['pct_counts_ribo'] <= 10]\n",
    "#Filter out cells with more than 15000 counts\n",
    "combined_anndata = combined_anndata[combined_anndata.obs['total_counts'] < 15000].copy()\n",
    "#Filter out cells with less than 200 genes counts\n",
    "sc.pp.filter_cells(combined_anndata, min_genes=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#PLOTTING THE QC METRICS\n",
    "\n",
    "p1 = sns.displot(combined_anndata.obs[\"total_counts\"], bins=100, kde=False)\n",
    "p2 = sc.pl.violin(combined_anndata, \"pct_counts_mt\")\n",
    "p2 = sc.pl.violin(combined_anndata, \"pct_counts_ribo\")\n",
    "p3 = sc.pl.scatter(combined_anndata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we annotate the cell types using Cell-typist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#We store the raw counts in the \"raw_counts\" layer of the combined_anndata object for future use\n",
    "combined_anndata.layers[\"raw_counts\"]=combined_anndata.X.copy()\n",
    "\n",
    "#Cell typist requires the data to be log normalized and scaled to a library size of 1e4\n",
    "\n",
    "#normalize the data and scale it to 10000\n",
    "sc.pp.normalize_total(combined_anndata, target_sum=1e4)\n",
    "#log transform the data\n",
    "sc.pp.log1p(combined_anndata)\n",
    "\n",
    "\n",
    "#predict the cell types\n",
    "cell_type_predictions = celltypist.annotate(combined_anndata, model = 'Lethal_COVID19_Lung.pkl', majority_voting = True)\n",
    "#transfer the predictions to the annotation data object\n",
    "predictions_adata = cell_type_predictions.to_adata()\n",
    "\n",
    "#add biological condition to the cells using the information from the sample name\n",
    "combined_anndata.obs['Condition'] = combined_anndata.obs['Sample'].apply(lambda x: 'covid' if x.endswith('cov') else 'control')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now separate the Epithelial cells for further investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Epithelial cell types In the Lung\n",
    "cell_types = [\"AT1\",\"AT2\",  \"Airway goblet\",\"Airway ciliated\",\"Airway basal\",\"Airway club\", 'Airway mucous', \"ECM-high epithelial\",\"Cycling epithelial\"]\n",
    "#Keep only these cell types and filter our the rest\n",
    "combined_anndata=combined_anndata[combined_anndata.obs[\"majority_voting\"].isin(cell_types)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLOT THE UMAP\n",
    "#Compute the UMAP\n",
    "sc.tl.umap(combined_anndata)\n",
    "#Plot the UMAP highlighting the cell types\n",
    "sc.pl.umap(combined_anndata, color=['majority_voting'], frameon=True, legend_fontsize=8)\n",
    "#Plot the UMAP highlighting the samples\n",
    "sc.pl.umap(combined_anndata, color=\"Sample\",frameon=True)\n",
    "sc.pl.umap(combined_anndata, color=\"Condition\",frameon=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we carry out differential expression analysis through pseudobulking and visualize the results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#We use raw counts for carrying out DGE\n",
    "combined_anndata.X=combined_anndata.layers[\"raw_counts\"].copy()\n",
    "\n",
    "#the genes of interst we want to investigate\n",
    "antiviral_zap=[\"ZC3HAV1\"]\n",
    "rna_editing_enzymes=['AICDA', 'APOBEC1', 'APOBEC2', 'APOBEC3A', 'APOBEC3B', 'APOBEC3C', 'APOBEC3D', 'APOBEC3F', 'APOBEC3G', 'APOBEC3H', 'APOBEC4','ADAR', 'ADARB1', 'ADARB2']\n",
    "gene_list=antiviral_zap+rna_editing_enzymes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to plot the volcano plot\n",
    "def plot_volcano(dea_results, genes_of_interest, logfc_threshold=.75, padj_threshold=0.1, adjust=True):\n",
    "    \"\"\"\n",
    "    Plot a volcano plot to visualize differential gene expression analysis results.\n",
    "\n",
    "    Parameters:\n",
    "    - dea_results (dataframe): The dataframe containing differential gene expression analysis results.\n",
    "    - genes_of_interest (list): A list of genes of interest to be highlighted in the plot.\n",
    "    - logfc_threshold (float, optional): The threshold for log2 fold change. Default is 1.\n",
    "    - padj_threshold (float, optional): The significance threshold for adjusted p-value. Default is 0.1.\n",
    "    - adjust (bool, optional): Whether to adjust the position of gene labels to avoid overlap. Default is True.\n",
    "\n",
    "    \"\"\"\n",
    "    #plot size\n",
    "    plt.figure(figsize=(10, 10))\n",
    "\n",
    "\n",
    "    # Filter the genes based on the thresholds\n",
    "    nonsig_genes = dea_results[(dea_results['padj'] > padj_threshold) | (abs(dea_results['log2FoldChange']) < logfc_threshold)]\n",
    "    sig_genes = dea_results[(dea_results['padj'] <= padj_threshold) & (abs(dea_results['log2FoldChange']) >= logfc_threshold)]\n",
    "\n",
    "    # Plot non-significant genes\n",
    "    plt.scatter(nonsig_genes['log2FoldChange'], -np.log10(nonsig_genes['padj']), c='#DCDCDC', edgecolor='#E0E0E0', label='Not Statistically Significant', marker='o')\n",
    "\n",
    "    # Plot significant genes\n",
    "    plt.scatter(sig_genes['log2FoldChange'], -np.log10(sig_genes['padj']), c='#A9A9A9', edgecolor='#E0E0E0', label='Statistically Significant', marker='o')\n",
    "\n",
    "    # Plot the genes of interest with a different marker\n",
    "    genes_of_interest_points = dea_results[dea_results['gene_symbol'].isin(genes_of_interest)]\n",
    "    plt.scatter(genes_of_interest_points['log2FoldChange'], -np.log10(genes_of_interest_points['padj']), c='b', edgecolor='none', label='Genes of Interest', marker='^')\n",
    "\n",
    "    # Draw a horizontal line for padj threshold\n",
    "    plt.axhline(y=-np.log10(padj_threshold), color='#909090', linestyle='--', linewidth=1)\n",
    "\n",
    "    # Draw two vertical lines for log2foldchange threshold\n",
    "    plt.axvline(x=logfc_threshold, color='#909090', linestyle='--', linewidth=1)\n",
    "    plt.axvline(x=-logfc_threshold, color='#909090', linestyle='--', linewidth=1)\n",
    "\n",
    "    # Label the genes of interest \n",
    "    texts = []\n",
    "    for gene in genes_of_interest:\n",
    "        gene_data = dea_results[dea_results['gene_symbol'] == gene]\n",
    "        if not gene_data.empty:\n",
    "            x = gene_data['log2FoldChange'].values[0]\n",
    "            y = -np.log10(gene_data['padj'].values[0])\n",
    "            txt = plt.text(x, y, gene, size=20, ha='center', va='center') \n",
    "            txt.set_path_effects([PathEffects.withStroke(linewidth=3, foreground='w')])\n",
    "            texts.append(txt)\n",
    "\n",
    "    plt.legend()\n",
    "    plt.xlabel(\"$log_{2}$ fold change\", size=15)\n",
    "    plt.ylabel(\"-$log_{10}$ $p_{adj}$-value\", size=15)\n",
    "\n",
    "    # Adjust text to avoid overlap\n",
    "    if adjust:\n",
    "        adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color='k', lw=0.5))\n",
    "\n",
    "    # Set the size of x-ticks and y-ticks\n",
    "    plt.xticks(size=12)\n",
    "    plt.yticks(size=12)\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Minimum number of cells per sample required to form pseudobulks\n",
    "minimum_cell_population_per_sample=30\n",
    "#Minimum number of pseudobulks per condition required to perform DGE analysis\n",
    "minimum_sample_per_condition=3\n",
    "\n",
    "#To record the log2foldchange data for each cell type\n",
    "lfc_data=[]\n",
    "#To record the cell types that qualified for pseudobulk-DGE analysis\n",
    "cell_studied=[]\n",
    "\n",
    "#iterate through each of the epithelial cell types\n",
    "for cell_type in cell_types:\n",
    "    #variables to record the number of control and covid pseudobulks\n",
    "    no_of_control_pseudobulks=0\n",
    "    no_of_covid_pseudobulks=0\n",
    "    samples=[]\n",
    "    #print the cell type under investigation\n",
    "    print(f\"Cell Type: {cell_type}\")\n",
    "    #To record the combined gene counts for the pseudobulks samples for this cell type\n",
    "    combined_count_2d=[]\n",
    "    #Subsetting the annotation data object for this cell type\n",
    "    cell_subset=[]\n",
    "    cell_subset = combined_anndata[combined_anndata.obs['majority_voting'] == cell_type]\n",
    "    \n",
    "    #Iterate through each of the samples(biological replicate) for this cell type to form pseudobulks\n",
    "    for sample in cell_subset.obs['Sample'].unique():\n",
    "        samp_cell_subset=[]\n",
    "        samp_cell_subset = cell_subset[cell_subset.obs['Sample'] == sample]\n",
    "\n",
    "        #check if the pseudobulk has the minimum cell population required to form a pseudobulk\n",
    "        if(len(samp_cell_subset)>=minimum_cell_population_per_sample):\n",
    "            print(f\"{sample} satisfies minimum cell population criteria: {len(samp_cell_subset.obs_names)}\")\n",
    "            #record the counts for the pseudobulk\n",
    "            total_counts = [int(count) for count in samp_cell_subset.X.sum(axis=0)]\n",
    "            #record the pseudobulk sample name and biological condition for generating the metadata\n",
    "            samples.append(sample)\n",
    "            combined_count_2d.append(total_counts)\n",
    "            if(sample[3:6] == \"ctr\"):\n",
    "                no_of_control_pseudobulks+=1\n",
    "            else:  \n",
    "                no_of_covid_pseudobulks+=1\n",
    "        else:\n",
    "            #remove the sample if it does not have the minimum number of cells required to form a pseudobulk\n",
    "            print(f\"{sample} does not satisfy minimum cell population criteria: {len(samp_cell_subset.obs_names)}\")\n",
    "    #check if we have the minimum number of pseudobulks per condition to perform DEA for this cell type   \n",
    "    if(no_of_control_pseudobulks>=minimum_sample_per_condition and no_of_covid_pseudobulks>=minimum_sample_per_condition):\n",
    "        #PERFORM DEA analysis using DESeq2\n",
    "        #create a counts dataframe using the combined gene counts from the pseudobulks\n",
    "        counts_df = pd.DataFrame(combined_count_2d, index=samples, columns = combined_anndata.var_names) \n",
    "        #create metadata\n",
    "        meta_data = pd.DataFrame([sample[3:6] for sample in samples], index=samples, columns = [\"Condition\"])\n",
    "        print(\"Counts DataFrame shape:\", counts_df.shape)\n",
    "        print(\"Metadata DataFrame shape:\", meta_data.shape)\n",
    "\n",
    "        #keep genes that have at least a total count of 5\n",
    "        genes_to_keep = counts_df.columns[counts_df.sum(axis=0) >= 5]\n",
    "        counts_df = counts_df[genes_to_keep]\n",
    "\n",
    "        #run DESeq2\n",
    "        dds = DeseqDataSet(\n",
    "            counts=counts_df,\n",
    "            metadata=meta_data,\n",
    "            design_factors=\"Condition\",\n",
    "            n_cpus=4, \n",
    "        )\n",
    "        dds.deseq2()\n",
    "        stat_ressults = DeseqStats(dds, contrast=('Condition', 'cov', 'ctr'))\n",
    "        stat_ressults.summary()\n",
    "\n",
    "        #Get the DEA analysis results\n",
    "        dea_results  = stat_ressults.results_df\n",
    "        dea_results[\"gene_symbol\"]=dea_results.index\n",
    "\n",
    "        # PLOT THE VOLCANO PLOT\n",
    "        plot_volcano(dea_results, gene_list)\n",
    "\n",
    "        #CARRY OUT GENE SET ENRICHMENT ANALYSIS\n",
    "        #rank the genes based on the product of the negative logarithm of the adjusted p-value and log2foldchange\n",
    "        dea_results['Rank'] = -np.log10(dea_results[\"padj\"])*dea_results[\"log2FoldChange\"]\n",
    "        ranking = dea_results[['Rank']]\n",
    "        ranking['Gene']= ranking.index\n",
    "        ranking = ranking.reset_index(drop=True)\n",
    "        ranking = ranking.reindex(columns=['Gene'] + [col for col in ranking.columns if col != 'Gene'])\n",
    "        #path to the gene set for oxidative damage response pathway\n",
    "        gene_set=\"WP_OXIDATIVE_DAMAGE_RESPONSE.v2023.2.Hs.gmt\"\n",
    "        #perform GSEA\n",
    "        pre_res = gp.prerank(rnk = ranking, gene_sets =  gene_set, seed = 6, min_size=3)\n",
    "        #plot the gene set enrichment analysis results\n",
    "        axs = pre_res.plot(terms=\"WP_OXIDATIVE_DAMAGE_RESPONSE\",figsize=(10,10))\n",
    "        \n",
    "\n",
    "        #RECORD THE LOG FOLD CHANGES OF THE GENES OF INTEREST\n",
    "        #extract the Differntial Expression Analysis results for the genes of interest\n",
    "        genes_of_interest_df=dea_results[dea_results.index.isin(gene_list)]\n",
    "        #reorder the index to match the order in the gene list\n",
    "        genes_of_interest_df=genes_of_interest_df.reindex(gene_list)\n",
    "        #record the log2foldchange for this cell type\n",
    "        lfc_data.append(genes_of_interest_df[\"log2FoldChange\"].to_list())\n",
    "        #record the cell type that was studied\n",
    "        cell_studied.append(cell_type)\n",
    "\n",
    "#PLOT THE HEATMAP LOG FOLD CHANGE OF THE GENES OF INTEREST ACROSS THE CELL TYPES\n",
    "lfc_df = pd.DataFrame(np.transpose(lfc_data), index=gene_list, columns=cell_studied)\n",
    "plt.figure(figsize=(4, 8))\n",
    "ax = sns.heatmap(lfc_df, annot=True, cmap='coolwarm', linewidths=.5, fmt='.2f')  \n",
    "# Rotate x ticks by 90 degrees\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=90, fontsize=12)\n",
    "# Decrease the overall font size\n",
    "sns.set_context(\"paper\", font_scale=1)  \n",
    "# Set the title\n",
    "ax.set_title(f\"Log Fold Change(Mean) in Cell types\", fontsize=20, fontweight='bold', pad=20) \n",
    "# Set the x-axis label\n",
    "ax.set(xlabel=\"Cells\", ylabel=\"Genes\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We save the Processes Annotated Data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the annotated adata object\n",
    "combined_anndata.write(\"processed_adata.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
